protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/fastfail/AbstractValidRcptHandler.java

```diff
@@ -31,58 +31,47 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * Handler which want todo an recipient check should extend this
- *
+ * Handler which want to do an recipient check should extend this
  */
 public abstract class AbstractValidRcptHandler implements RcptHook {
     private static final Logger LOGGER = LoggerFactory.getLogger(AbstractValidRcptHandler.class);
 
     @Override
     public HookResult doRcpt(SMTPSession session, MailAddress sender, MailAddress rcpt) {
-        boolean reject = false;
-        
-        if (session.isRelayingAllowed()) {
-            // check if the domain is local, if so we still want to check if the recipient is valid or not as we want to fail fast in such cases
-            if (isLocalDomain(session, rcpt.getDomain())) {
-                if (isValidRecipient(session, rcpt) == false) {
-                    reject = true;
-                }
-            }
-        } else {
-            if (isLocalDomain(session, rcpt.getDomain()) == false) {
-                LOGGER.debug("Unknown domain {} so reject it", rcpt.getDomain());
+        if (!isLocalDomain(session, rcpt.getDomain())) {
+            return handleRemoteDomain(session, rcpt);
+        }
+        return handleLocalDomain(session, rcpt);
+    }
 
-            } else {
-                if (isValidRecipient(session, rcpt) == false) {
-                    reject = true;
-                }
-            }
+    public HookResult handleLocalDomain(SMTPSession session, MailAddress rcpt) {
+        if (!isValidRecipient(session, rcpt)) {
+            return reject(rcpt);
         }
-       
-        if (reject) {
-            //user not exist
-            LOGGER.info("Rejected message. Unknown user: {}", rcpt);
-            return new HookResult(HookReturnCode.DENY,SMTPRetCode.MAILBOX_PERM_UNAVAILABLE, DSNStatus.getStatus(DSNStatus.PERMANENT,DSNStatus.ADDRESS_MAILBOX) + " Unknown user: " + rcpt.toString());
-        } else {
-            return HookResult.declined();
+        return HookResult.declined();
+    }
+
+    public HookResult handleRemoteDomain(SMTPSession session, MailAddress rcpt) {
+        if (!session.isRelayingAllowed()) {
+            LOGGER.debug("Unknown domain {} so reject it", rcpt.getDomain());
```

Comment:
you don't return anything here, is it intended ?

Reply:
Intended. It's a refactoring hence I did not introduced behaviour changes.

BTW if I do, then the error upon SMTP relay of unauthenticated users is `User do not exist` instead of `Relay denied`. 

Which is legitimate: the local server has not idea if the distant recipient exists or not.

However I could be removing this misleading piece of log in a separate commit. (WHich would further make the code orthogonal to relay logic).